common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,208 @@
+            recipient.cancel(false);
+        } else {
+            Throwable cause = completed.cause();
+            tryFailure(recipient, cause, LOGGER);
+        }
+    }
+
+    private static class PassThrough<R> implements FutureContextListener<Promise<R>, Object> {
+        @Override
+        public void operationComplete(Promise<R> recipient, Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    R result = (R) completed.getNow();
+                    recipient.trySuccess(result);
+                } catch (RuntimeException e) {
+                    tryFailure(recipient, e, LOGGER);
+                }
+            } else {
+                propagateUncommonCompletion(completed, recipient);
+            }
+        }
+    }
+
+    private static class CallableMapper<R, T> implements Callable<R> {
+        private final Future<T> future;
+        private final Function<T, R> mapper;
+
+        CallableMapper(Future<T> future, Function<T, R> mapper) {
+            this.future = future;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public R call() throws Exception {
+            return mapper.apply(future.getNow());
+        }
+    }
+
+    private static class Mapper<R, T> implements FutureListener<Object> {
+        private final Promise<R> recipient;
+        private final Function<T, R> mapper;
+
+        Mapper(Promise<R> recipient, Function<T, R> mapper) {
+            this.recipient = recipient;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public void operationComplete(Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    T result = (T) completed.getNow();
+                    R mapped = mapper.apply(result);
+                    recipient.trySuccess(mapped);
+                } catch (RuntimeException e) {
+                    tryFailure(recipient, e, LOGGER);
+                }
+            } else {
+                propagateUncommonCompletion(completed, recipient);
+            }
+        }
+    }
+
+    private static class FlatMapper<R, T> implements FutureListener<Object> {
+        private final Promise<R> recipient;
+        private final Function<T, Future<R>> mapper;
+
+        FlatMapper(Promise<R> recipient, Function<T, Future<R>> mapper) {
+            this.recipient = recipient;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public void operationComplete(Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    T result = (T) completed.getNow();
+                    Future<R> future = mapper.apply(result);
+                    if (future.isSuccess()) {
```

Comment:
nit: I find the hidden intent of `isSuccess()` inherently behaving like `isDone()` elsewhere, hard to parse. It confuses me as to why for `future.isDone() && !future.isSuccess()` we go through the listener path when we can simply terminate the promise.

I think we either do the bypass for the "future already done" case or we don't, this middle path is confusing IMO.

Reply:
Added shortcuts for failed futures as well.